provisioner/shell/provisioner_test.go

```diff
@@ -98,6 +98,50 @@ func TestProvisionerPrepare_InlineShebang(t *testing.T) {
 	}
 }
 
+func TestProvisionerPrepare_ShebangWithinInline(t *testing.T) {
+	config := testConfig()
+	script := `#!/bin/bash
+		echo "Hello, World!"
+	`
+	config["inline"] = []interface{}{script, "foo", "bar"}
+
+	p := new(Provisioner)
+	err := p.Prepare(config)
+	if err != nil {
+		t.Fatalf("should not have error: %s", err)
+	}
+
+	if p.config.InlineShebang != "/bin/bash" {
+		t.Fatalf("bad value: %s", p.config.InlineShebang)
+	}
+
+	// Test with InlineShebang, it should override the shebang in the script
+	config["inline_shebang"] = "foo"
+	p = new(Provisioner)
+	err = p.Prepare(config)
+	if err != nil {
+		t.Fatalf("should not have error: %s", err)
+	}
+
+	if p.config.InlineShebang != "foo" {
+		t.Fatalf("bad value: %s", p.config.InlineShebang)
+	}
+
+	// Test with a inline that does not have a shebang
+	config["inline"] = []interface{}{"foo", "bar"}
+	delete(config, "inline_shebang")
+	p = new(Provisioner)
```

Comment:
Given that the logic is duplicated here, I would suggest you run those as subtests, something like https://github.com/hashicorp/packer/blob/main/hcl2template/function/datetime_test.go#L35
Although the other tests in this file seem to adopt the same pattern as yours, so we could also keep it that way and change it later